### PR TITLE
more efficient MDC in Typed persistence, #24787

### PR DESCRIPTION
**Pull request by patriknw**
The original issue was solved in #24729 but here are some improvements to MDC in the persistence impl.

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala [hunk 1]
```diff
@@ -282,20 +282,6 @@ object Behaviors {
...
-   *                 each message processing by the inner behavior is done.
-   * @param behavior The actual behavior handling the messages, the MDC is used for the log entries logged through
-   *                 `ActorContext.log`
-   *
-   * See also [[akka.actor.typed.Logger.withMdc]]
-   */
-  def withMdc[T](
-    mdcForMessage: akka.japi.function.Function[T, java.util.Map[String, Any]], behavior: Behavior[T]): Behavior[T] =
-    WithMdcBehavior[T](Map.empty, message ⇒ mdcForMessage.apply(message).asScala.toMap, behavior)
-
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: I thought the 3 overloads were overkill, and they are bad for type inference (and compilation error messages).

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala [hunk 2]
```diff
@@ -61,5 +61,24 @@ private[persistence] final case class EventsourcedSetup[C, E, S](
 
   def selfUntyped = context.self.toUntyped
 
+  private var mdc: Map[String, Any] = Map.empty
+  private var _log = context.log // changed when mdc is changed
+  def log: Logger = {
+    if (_log == null)
```
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: `OptionVal` > `null`

Also, is this optimization really needed, putting a lazy var depending on another var inside a case class smells a bit to me. How many transitions/new mdc decorators was created before?
Could we do the same optimization in some other less imperative way? Just add a field to the case class?

Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: OptionVal no prob.

We switch mdc for each persisted event. Logging is rare so creating new Logger for each switch felt wasteful.

Copy of setup for each switch also felt wasteful and would introduce a var in other place. Setup is anyway not fully immutable.

Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: Hmm. ok, fair enough. Thanks for the motivation.

Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: Should we maybe just make it a regular class altogether then? Avoiding any copy at all...

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala (hunk 1 above)
Comment (ktoso) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: I see that you mean with the overloads but I think they all pull their weight hm

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala [hunk 3]
```diff
@@ -315,6 +301,8 @@ object Behaviors {
    * are in both the static and the per message MDC the per message one overwrites the static one
    * in the resulting log entries.
    *
+   * * The `staticMdc` or `mdcForMessage` may be empty or `null`.
```
[the remark this answers is not in the document]
Comment (ktoso) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: +1

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala [hunk 4]
```diff
@@ -29,19 +29,20 @@ private[akka] object EventsourcedBehavior {
...
-      "persistenceId" → setup.persistenceId,
-      "phase" → phaseName
-    ))(b)
+    def create(persistenceId: String, phaseName: String): Map[String, Any] = {
+      Map(
+        "persistenceId" → persistenceId,
+        "phase" → phaseName
+      )
+    }
+  }
```
Comment (ktoso) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala]: ok

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala [hunk 5]
```diff
@@ -61,5 +61,24 @@ private[persistence] final case class EventsourcedSetup[C, E, S](
...
+  def setMdc(newMdc: Map[String, Any]): EventsourcedSetup[C, E, S] = {
+    mdc = newMdc
+    _log = null // lazy init
+    this
+  }
+
+  def setMdc(phaseName: String): EventsourcedSetup[C, E, S] = {
+    setMdc(MDC.create(persistenceId, phaseName))
+    this
+  }
```
Comment (ktoso) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: ok

**Review by ktoso — APPROVED** [review]
LGTM though I would keep the overload,

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala (hunk 1 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: To clarify, the problem is for example:

```
      def behavior3: Behavior[String] =
        Behaviors.withMdc[String](msg ⇒ Map("a" -> msg)) {
          Behaviors.receive {
            case (ctx, msg) ⇒
              ctx.log.info(msg.toUpperCase())
              Behaviors.same
          }
        }

[error] /Users/patrik/dev/akka/akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:323:27: missing parameter type
[error]         Behaviors.withMdc(msg ⇒ Map("a" -> msg)) {
```

One would have to add `(msg: String) =>`

It only goes away completely if removing all overloads and only keeping
```
def withMdc[T](staticMdc: Map[String, Any], mdcForMessage: T ⇒ Map[String, Any])(behavior: Behavior[T]): Behavior[T]
```

So my removal of one overloads is not worth much, reverting it and creating a separate ticket.

Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: I played around with giving the static one a separate name `withStaticMdc` earlier today but didn't see any improvement, but maybe that is the path to go down three separate names?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala (hunk 2 above)
Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: Doesn't necessarily have to be done in this PR though.

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala (hunk 1 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: yeah, having separate names would solve it (then not overloading), however how big api surface do we really need for mdc? Then it should start with same `withMdc...` but I can't come up with any good such naming

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala (hunk 2 above)
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedSetup.scala]: changed

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala [hunk 6]
```diff
@@ -29,19 +29,20 @@ private[akka] object EventsourcedBehavior {
...
-    val ReplayingSnapshot    = "replay-snap"
-    val ReplayingEvents      = "replay-evts"
-    val RunningCmds           = "running-cmnds"
-    val PersistingEvents      = "persist-evts"
+    val AwaitingPermit = "get-permit"
+    val ReplayingSnapshot = "replay-snap"
+    val ReplayingEvents = "replay-evts"
+    val RunningCmds = "running-cmnds"
+    val PersistingEvents = "persist-evts"
     // format: ON
```
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedBehavior.scala]: reverted, I don't know

**Comment by patriknw** [comment]
incorporated feedback, see 270d4a8

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: `withMdcStatic`, `withMdcPerMessage`, `withMdc` doesn't sound so bad or crazy api-surface-wise to me.

**Review by johanandren — APPROVED** [review]
LGTM, not sure how important the mdc-discussion is.

File: akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala (hunk 1 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala]: Created issue https://github.com/akka/akka/issues/24798, with another suggestion

**Pull request merged**